### ISSUE TITLE
Support generating random 64-bit values.

### ISSUE THREAD

**Issue by tannergooding** [dotnet/corefx#30933]
### Rationale

It is sometimes desirable to generate random numbers as inputs to tests. This works nicely for the majority of the primitive types, but not for `long`/`ulong`.

As such, I propose a method be exposed that allows the generation of random 64-bit integer values.

### Proposed API
```C#
public class Random
{
    public virtual int NextInt64();
    public virtual int NextInt64(long maxValue);
    public virtual int NextInt64(long minValue, long maxValue);
}
```

**Comment by Tornhoof** [comment]
Shouldn't be the return value be long then?

**Comment by tannergooding** [comment]
> Shouldn't be the return value be long then?

Yep. Fixed

**Comment by stephentoub** [comment]
In general I think such an addition makes sense, but some questions:
- Would all three overloads actually be virtual?
- How do you see the base implementations being implemented?  In particular, if someone's derived from Random and overridden other virtuals, presumably those virtuals should be used (otherwise the new methods could utilize the wrong algorithm)... which methods would these delegate to?

**Comment by tannergooding** [comment]
@stephentoub, the short answer to your questions is that they are virtual because the existing `Random.Next()` methods are virtual and they should behave/be implemented following the same rules as the existing algorithms (which is, overloading one does not impact the other two).

This follows the "existing" practice for the `System.Random` class, which AFAIK is preferred.

**Comment by stephentoub** [comment]
What about my second question?

**Comment by tannergooding** [comment]
I don't think they should/can reliably delegate. The overall range of the existing methods isn't sufficient to provide good distribution for `long`/`ulong` (unless we default the range to the same as `int`).

If a user has already extended `System.Random` and overridden `Next`, they would likely need to reimplement `NextInt64` as well. Thinking about it, they could also have already exposed a `NextInt64` method, in which case this could be a "breaking" change for recompilation where warn as error is enabled.

An alternative would be to extend `System.Random` ourselves and not make these methods virtual at all.

**Comment by stephentoub** [comment]
> I don't think they should/can reliably delegate.

That's problematic then (it's why I was asking).  Let's say I've created my own `SecureRandom : Random` type that uses `RandomNumberGenerator` under the covers, and I've overridden all of the virtuals to behave the way I need them to.  Now we add this new base virtual method, but it's not implemented in terms of any of the existing virtuals.  Someone starts using my type before I've had a chance to provide a new override, and they call the new methods, and they get functionality that behaves incorrectly.

> Thinking about it, they could also have already exposed a NextInt64 method, in which case this could be a "breaking" change for recompilation where warn as error is enabled.

While true, that's a problem we face any time we add anything new... a new type could conflict with a type the customer named, a new method on anything that's not sealed could conflict with a method the customer added on a derived type, a new method could start to be targeted silently instead of an extension method of the same name the customer defined, etc.

**Comment by tannergooding** [comment]
> That's problematic then (it's why I was asking).

Right. But I don't think we can provide new overloads, short of implementing our own sub-class/providing a new Random type (or providing some kind of compat-switch), otherwise. But maybe that is a direction worth taking...

`System.Random` is already fairly "frozen-in-time". AFAIK, we can't readily change the algorithm (as it is a massive breaking change to anyone relying on the deterministic output from providing a fixed-seed) and because it isn't sealed it is hard to add new overloads/methods.

**Comment by danmosemsft** [comment]
Can't the default implementations just call the existing int implementations twice (and shift one), to create an appropriate long?

**Comment by tannergooding** [comment]
> Can't the default implementations just call the existing int implementations twice (and shift one), to create an appropriate long?

Yes, but that doesn't necessarily provide good distribution.

**Comment by stephentoub** [comment]
> But maybe that is a direction worth taking...

For just this method?  Meh, I don't think so.  There must be ways of getting the desired distribution from one of the existing sources of randomness, whether Sample(), Next(...), NextBytes, NextDouble, etc., or some combination of them, even if it makes the NextInt64 method more expensive than if it wasn't delegating.

**Comment by danmosemsft** [comment]
> Yes, but that doesn't necessarily provide good distribution.

It does if the existing API is producing a good distribution (mixing random bits gives random bits)?

> It would also make the algorithm more complex in the scenarios where the user specifies a custom range.

Yes, but seems doable.

**Comment by tannergooding** [comment]
> It does if the existing API is producing a good distribution (mixing random bits gives random bits)?

Right, but because these are pseudo-random number generation algorithms they don't have perfect distribution and that can impact the overall set of numbers generated.

Finding out the distribution characteristics of the existing algorithm (when applied to generating `Int64`) would be something that needs to be done as part of the implementation process.

**Comment by morganbr** [comment]
@tannergooding, are you saying that there's more of a relationship between two consecutive outputs from System.Random than there would be from the same algorithm extended to cover 64-bit numbers? I'd be very surprised if that was the case. One could make a case that the actual algorithm in System.Random is suboptimal in general, but I don't think the size of the output is important in that.

**Comment by jnm2** [comment]
I'm just curious how you would compose this. Let's say you wanted to generate a random number between 0 and 0x1_0000_0000, inclusive. If you generate the high 32 bits separately, from the low 32 bits, and you use `random.Next(0, 2)` to do so, there is a 50% chance that the final 64 bit number will be 0x1_0000_0000 and a 50% change that it will be any of 4 billion other numbers.

**Comment by jnm2** [comment]
I'm just curious how you would compose this. Let's say you wanted to generate a random number between 0 and 0x1_0000_0000, inclusive. If you generate the high 32 bits separately from the low 32 bits, and you use `random.Next(0, 2)` to do so, there is a 50% chance that the final 64 bit number will be 0x1_0000_0000 and a 50% chance that it will be any of 4 billion other numbers.

**Comment by tannergooding** [comment]
> are you saying that there's more of a relationship between two consecutive outputs from System.Random than there would be from the same algorithm extended to cover 64-bit numbers

I am saying that any inefficiencies or limitations of the current algorithm will be at risk of becoming more apparent.

Since the algorithm was designed for generating integers in the range of 0 to 2.14b, numbers outside that range are no longer guaranteed to be uniform (occur once in a cycle) and the cycle length will be halved (as we require at least two samples per generated number -- we already require 2 samples for negative 32-bit integers, and some 64-bit numbers will require 3 samples).

Additionally, our algorithm deviates from the "Subtractive Number Generator" given by Donald Knuth (https://github.com/dotnet/corefx/issues/23298), so we don't have the same guarantees of uniformity, cycle length, etc that we could potentially rely on to generate these numbers.

Our own documentation touches briefly on how to use `System.Random` to generate 64-bit values, and some of the downsides of doing so: https://docs.microsoft.com/en-us/dotnet/api/system.random?view=netcore-2.1#generate-random-64-bit-integers

**Comment by colgreen** [comment]
Bear in mind (as per @tannergooding's comments) that System.Random has a number of issues (see #23298 for details) and thus probably should not form the basis of any work in the core product requiring a source of pseudo randomness.

For a PRNG implementation that addresses all of the known issues, see:

https://github.com/colgreen/Redzen/blob/master/Redzen/Random/Xoshiro256PlusRandom.cs

or (same algorithm with more PRNG state):

https://github.com/colgreen/Redzen/blob/master/Redzen/Random/Xoshiro512StarStarRandom.cs

I understand the point about not wanting to change System.Random as there will be code that relies on it's specific (and often quirky) behaviour, but it does have issues and ideally these would be addressed, perhaps as an alternative PRNG class.

Regarding the xoshiro PRNG versus the one in System.Random; there may be aspects of xoshiro that you find undesirable, however, some of the issues In System.Random are in how the random bits are utilised not the PRNG itself, and for those issues I believe the linked code above corrects all of them. (see #23298 and also the comments in the xoshiro source code).
